### PR TITLE
isbn-verifier: Include more succinct description of ISBN-10.

### LINKED ISSUE #1011: isbn-verifier: canonical-data.json (order and descriptions)
Hi all,

I have just gone about implementing `isbn-verifier` as part of the Java stream, and while doing so I noticed that the descriptions in `canonical-data.json` seem to be inconsistent (both in verbosity and syntax), plus I was not sure if there might be a better order for the tests.

Of course, it could have just been my own lack of understanding that made it difficult, but I did find implementing them confusing.

Does anyone here have any opinions on it? I'd like to hear :slightly_smiling_face:  

Update:
[https://github.com/exercism/problem-specifications/blob/master/exercises/isbn-verifier/canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/isbn-verifier/canonical-data.json)

### PR DESCRIPTION
**Pull request by sjwarner-bp**
Fixes #1011 .

### CODE REVIEW

File: exercises/isbn-verifier/description.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
-Check if a given ISBN-10 is valid.
+Given a number, check if a given ISBN-10 is valid.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: I'm not sure this change is an improvement.

Comment (Insti) [reply exercises/isbn-verifier/description.md]: Actually the whole line can probably be removed.

File: exercises/isbn-verifier/description.md [hunk 2]
```diff
@@ -9,27 +12,35 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
...
+
+```
+(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0
+```
+
+If the result is 0, then it is a valid ISBN-10, otherwise it is invalid.
+
+## Example
+
+Let's take a random ISBN-10 number, for instance `3-598-21508-8`.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: It's not really a random ISBN-10 number.
This line is unnecessarily verbose.

File: exercises/isbn-verifier/description.md (hunk 1 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Sure thing, I was just trying to mirror the [Luhn description](https://github.com/exercism/problem-specifications/blob/master/exercises/luhn/description.md), as I find it a really good example for this kind of problem.

File: exercises/isbn-verifier/description.md [hunk 3]
```diff
@@ -9,27 +12,35 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
...
 The first digit block indicates the group where the ISBN belongs. Groups can consist of shared languages, geographic regions or countries. The leading '3' signals this ISBN is from a german speaking country.
 The following number block is to identify the publisher. Since this is a three digit publisher number there is a 5 digit title number for this book.
 The last digit in the ISBN is the check digit which is used to detect read errors.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: All this stuff about countries and publishers doesn't belong in the example.

File: exercises/isbn-verifier/description.md (hunk 2 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: I agree, that's why I slimmed it down. Do you think it should be reduced even more?

File: exercises/isbn-verifier/description.md (hunk 3 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Again, I agree but was not confident in stripping it out without another opinion. If you think this should be removed then I'll do that too! :slightly_smiling_face:

File: exercises/isbn-verifier/description.md [hunk 4]
```diff
@@ -1,4 +1,7 @@
-Check if a given ISBN-10 is valid.
+Given a number, check if a given ISBN-10 is valid.
+
+The [ISBN-10 verification process](http://www.hahnlibrary.net/libraries/isbncalc.html) requires use of a simple formula, and is mostly used to validate book identification
+numbers.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: I'd re-phrase this, to focus on what the purpose of ISBNs are before talking about formulas.

File: exercises/isbn-verifier/description.md [hunk 5]
```diff
@@ -9,27 +12,35 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
...
 
-Let's take a random ISBN-10 number, say `3-598-21508-8` for this.
+The ISBN-10 format is 9 digits (0 to 9) plus one check character (either a digit or an X only). In the case the check character is an X, this represents the value '10'. These may be communicated with or without hyphens, and can be checked for their validity by the following formula:
+
+```
+(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0
+```
+
+If the result is 0, then it is a valid ISBN-10, otherwise it is invalid.
+
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: This is a good description 👍

File: exercises/isbn-verifier/description.md (hunk 4 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Ok, I'll give the problem context first :+1:

File: exercises/isbn-verifier/description.md [hunk 6]
```diff
@@ -9,27 +12,35 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
...
-The check digit can additionally be an 'X' to allow 10 to be a valid check digit as well.
-
-A valid ISBN-10 is calculated with this formula `(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0`
 So for our example ISBN this means:
+```
 (3 * 10 + 5 * 9 + 9 * 8 + 8 * 7 + 2 * 6 + 1 * 5 + 5 * 4 + 0 * 3 + 8 * 2 + 8 * 1) mod 11 = 0
+```
 
-Which proves that the ISBN is valid.
+Since the result is 0, this proves that our ISBN is valid.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Good clarification 👍

File: exercises/isbn-verifier/description.md (hunk 2 above)
Comment (Insti) [reply exercises/isbn-verifier/description.md]: Yes.
"For the ISBN-10 3-598..."

Be careful using "ISBN Number" since it has the "ATM Machine" issue.

Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: RAS Syndrome! Good point - I'll strip that too.

File: exercises/isbn-verifier/description.md [hunk 7]
```diff
@@ -1,4 +1,7 @@
-Check if a given ISBN-10 is valid.
+Given a number, check if a given ISBN-10 is valid.
+
+The [ISBN-10 verification process](http://www.hahnlibrary.net/libraries/isbncalc.html) requires use of a simple formula, and is mostly used to validate book identification
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Link to wikipedia and solve all your describe 'ISBN' problems: https://en.wikipedia.org/wiki/International_Standard_Book_Number

File: exercises/isbn-verifier/description.md (hunk 3 above)
Comment (Insti) [reply exercises/isbn-verifier/description.md]: The actual details of each section aren't relevant to the problem so removing them completely is not a bad option.

Another option is moving them up into the "about ISBN" section if we want to do more than just refer people to Wikipedia.

File: exercises/isbn-verifier/description.md [hunk 8]
```diff
@@ -9,27 +10,30 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
...
 
-Which proves that the ISBN is valid.
+Since the result is 0, this proves that our ISBN is valid.
 
 ## Caveats
 
-Converting from string to number can be tricky in certain languages.
-It's getting even trickier since the check-digit of an ISBN-10 can be 'X'.
+Converting from strings to numbers can be tricky in certain languages.
+Now, it's even trickier since the check digit of an ISBN-10 may be 'X' (representative of '10').
```
Comment (coriolinus) [inline exercises/isbn-verifier/description.md]: `representing` reads more cleanly to me than `representative of` here.

File: exercises/isbn-verifier/description.md [hunk 9]
```diff
@@ -1,4 +1,5 @@
-Check if a given ISBN-10 is valid.
+The [ISBN-10 verification process](http://www.hahnlibrary.net/libraries/isbncalc.html) is mostly used to validate book identification
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: I'd remove this link, and replace it with the wikipedia one. This link doesn't give much useful information about the number and the formula given there conflicts with the one specified here. (1-9) are reversed on that page.

Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Ah, I didn't notice that! On it.

File: exercises/isbn-verifier/description.md [hunk 10]
```diff
@@ -1,4 +1,5 @@
-Check if a given ISBN-10 is valid.
+The [ISBN-10 verification process](https://en.wikipedia.org/wiki/International_Standard_Book_Number) is mostly used to validate book identification
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: "mostly"?

File: exercises/isbn-verifier/description.md [hunk 11]
```diff
@@ -9,27 +10,30 @@ The program should allow for ISBN-10 without the separating dashes to be verifie
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: While we're here: Line 8 is grammatically odd.
Also this is the first indication that there could be dashes in the number.
Maybe add an example number to the first paragraph.

File: exercises/isbn-verifier/description.md [hunk 12]
```diff
@@ -1,4 +1,5 @@
-Check if a given ISBN-10 is valid.
+The [ISBN-10 verification process](https://en.wikipedia.org/wiki/International_Standard_Book_Number) is used to validate book identification
+numbers.
 
 ## Functionality
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Do any other `description.md`s have a Functionality section?

File: exercises/isbn-verifier/description.md (hunk 11 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Done :+1:

File: exercises/isbn-verifier/description.md (hunk 12 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: I thought this was odd too. I am not too sure what the original author means - can you think of a better title? :slightly_smiling_face:

File: exercises/isbn-verifier/description.md [hunk 13]
```diff
@@ -1,35 +1,39 @@
...
 (3 * 10 + 5 * 9 + 9 * 8 + 8 * 7 + 2 * 6 + 1 * 5 + 5 * 4 + 0 * 3 + 8 * 2 + 8 * 1) mod 11 = 0
+```
 
-Which proves that the ISBN is valid.
+Since the result is 0, this proves that our ISBN is valid.
 
 ## Caveats
 
-Converting from string to number can be tricky in certain languages.
-It's getting even trickier since the check-digit of an ISBN-10 can be 'X'.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Add an example valid ISBN with X as a check digit (from the canonical-data.json)
No need to show the calculation.

Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Shall do!

File: exercises/isbn-verifier/description.md (hunk 12 above)
Comment (Insti) [reply exercises/isbn-verifier/description.md]: What about: Move the whole functionality section under the example and rename it to 'Task'?

File: exercises/isbn-verifier/description.md [hunk 14]
```diff
@@ -30,7 +30,7 @@ Since the result is 0, this proves that our ISBN is valid.
 ## Caveats
 
 Converting from strings to numbers can be tricky in certain languages.
-Now, it's even trickier since the check digit of an ISBN-10 may be 'X' (representing '10').
+Now, it's even trickier since the check digit of an ISBN-10 may be 'X' (representing '10'). For instance `3-598-21507-X` is a valid ISBN-10.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: 👍

File: exercises/isbn-verifier/description.md [hunk 15]
```diff
@@ -1,35 +1,40 @@
...
+```
+
+Since the result is 0, this proves that our ISBN is valid.
+
+## Test
+
+Given an unknown string the program should check if the provided string is a valid ISBN-10.
+Putting this into place requires some thinking about preprocessing/parsing of the string prior to calculating the check digit for the ISBN.
+
+The program should allow for ISBN-10 both with and without separating dashes to be verified. For example, `3-598-21508-8` and `3598215088`.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: > The program should allow for ISBN-10 both with and without separating dashes to be verified. ~~For example, `3-598-21508-8` and `3598215088`.~~

File: exercises/isbn-verifier/description.md [hunk 16]
```diff
@@ -1,35 +1,40 @@
-Check if a given ISBN-10 is valid.
+The [ISBN-10 verification process](https://en.wikipedia.org/wiki/International_Standard_Book_Number) is used to validate book identification
+numbers.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Add: 
"These normally contain dashes and look like: `3-598-21508-8`"

**Comment by Insti** [comment]
Looks good, I need to go do some other stuff now.
I'll leave it a while and see what others have to suggest.
Thanks for all your work on this.

File: exercises/isbn-verifier/description.md [hunk 17]
```diff
@@ -1,35 +1,40 @@
...
-A valid ISBN-10 is calculated with this formula `(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0`
-So for our example ISBN this means:
+Let's take the ISBN-10 `3-598-21508-8`. We plug it in to the formula, and get:
+```
 (3 * 10 + 5 * 9 + 9 * 8 + 8 * 7 + 2 * 6 + 1 * 5 + 5 * 4 + 0 * 3 + 8 * 2 + 8 * 1) mod 11 = 0
+```
+
+Since the result is 0, this proves that our ISBN is valid.
+
+## Test
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Task?

File: exercises/isbn-verifier/description.md [hunk 18]
```diff
@@ -1,35 +1,40 @@
...
+
+Given an unknown string the program should check if the provided string is a valid ISBN-10.
```
Comment (Insti) [inline exercises/isbn-verifier/description.md]: Given a string, check if it is a valid ISBN-10

File: exercises/isbn-verifier/description.md (hunk 17 above)
Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Oops!

**Comment by Insti** [comment]
Closing github window :grin:

File: exercises/isbn-verifier/description.md (hunk 6 above)
Comment (rpottsoh) [reply exercises/isbn-verifier/description.md]: Should there be one `=` or two at the end of the line?  Line 18 only has one.

Comment (rpottsoh) [reply exercises/isbn-verifier/description.md]: See previous comment re: line 9.

**Comment by sjwarner-bp** [comment]
Hmm, that is a good point. Obviously this wants to be as language agnostic as possible - do you think '==' or '=' is better?

**Comment by coriolinus** [comment]
I'd prefer `==`.

**Comment by sjwarner-bp** [comment]
I'm inclined to agree with @coriolinus

File: exercises/isbn-verifier/description.md [hunk 19]
```diff
@@ -1,35 +1,40 @@
...
+
+## Task
+
+Given a string the program should check if the provided string is a valid ISBN-10.
+Putting this into place requires some thinking about preprocessing/parsing of the string prior to calculating the check digit for the ISBN.
 
-A valid ISBN-10 is calculated with this formula `(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0`
-So for our example ISBN this means:
-(3 * 10 + 5 * 9 + 9 * 8 + 8 * 7 + 2 * 6 + 1 * 5 + 5 * 4 + 0 * 3 + 8 * 2 + 8 * 1) mod 11 = 0
+The program should allow for ISBN-10 both with and without separating dashes to be verified.
```
Comment (ErikSchierboom) [inline exercises/isbn-verifier/description.md]: I think I would rephrase this slightly to emphasize that both are correct (the current wording only implies allowing them to be verified). Maybe something like:

`The program should be able to verify ISBN-10 both with and without separating dashes.`

Comment (sjwarner-bp) [reply exercises/isbn-verifier/description.md]: Done! :+1:

**Comment by rpottsoh** [comment]
Thanks for making this a better exercise @sjwarner-bp!
